Bespoke Halfpenny

swashcap opened the pull request:
* **Problem:** Generating Halfpenny via steelpenny makes for painful development because the code generation is difficult to reason about. It’s difficult to add features to the resulting client.
* **Solution:**
  * Make Halfpenny _custom, hand-coded_
  * Add Halfpenny tests
  * Update API client usage in steelpenny integration tests
  * Remove codegen apparatus
* **Testing:** `npm test` in _halfpenny_ and _steelpenny_.

### ToDos

- [x] Add documentation to Halfpenny
- [x] Update steelpenny documentation
- [x] Revert #91

jwtlake commented on packages/halfpenny/lib/halfpenny.js:
```diff
@@ -0,0 +1,638 @@
+    }
+
+    // Clear auth cookie and stored credentials
+    const removeAuth = (...args) => {
```
Did we have to do all this stuff before? I wasn't able to find this logic in `master`. Seems to be legacy support but I'm not totally sure what all is happening here.

swashcap replied:
There was a case cookies weren’t properly cleared in the COINS ecosystem. Perhaps we make authentication storage occur via an adapter in the future so it’s environment aware (Node.js versus a browser).

swashcap replied:
Some background: part of this comes from MRN-Code/nodeapi before we moved to a generated API client. Another part comes from the browser-centric authentication module in MRN-Code/coins-logon-widget. The most recent part comes from MRN-Code/coins-pr-site’s  API client. Some code/decisions are informed by the code’s past as well as COINS architecture.

swashcap commented on packages/steelpenny/test/integration/keys.js:
```diff
@@ -30,7 +30,10 @@ describe('POST keys (login)', () => {
       const handleAuthError = err => err;
 
-      return apiClient.auth.login('wrongUser', 'wrongPwd')
+      return apiClient.login({
```
It’s easier to manage new arguments if everything’s rolled into a single object parameter. I suppose this isn’t necessary for login, but I think every public method adheres to this. It’s consistent.

swashcap commented on packages/halfpenny/lib/halfpenny.js:
```diff
@@ -0,0 +1,638 @@
+   * @returns {Object}
+   */
+  removeCredentials() {
+    return this.setCredentials({
```
Good point. This looks like a holdover from [when we included nodeapi’s test auth module in _coins-logon-widget_](https://github.com/MRN-Code/nodeapi/blob/1bc9766038cace231fa4028eecde276743e8d1e5/src/test/sdk/authentication.js#L108). I’ll add a test and drop this.

swashcap replied:
Fixed in 483c9453d176dd3957fd32f85aa64bb423fec020 (with a test!)

swashcap commented on packages/halfpenny/lib/halfpenny.js:
```diff
@@ -0,0 +1,638 @@
+      return Promise.reject(new Error('No email or username'));
+    }
+
+    return this.get(`/users/check?${queryParams.join('&')}`)
```
That’s correct. I’m thinking this is easier:

```js
hp.checkUser({ email: 'moustaches@mrn.org' })
  .then((response) => {
    if (response.data) {
      console.log('User already exists');
    } else {
      console.log('User doesn\'t exist');
    }
  });
```

…versus that long, weird lookup:

```js
hp.checkUser({ email: 'moustaches@mrn.org' })
  .then((response) => {
    if (response.data.data[0].exists) {
      console.log('User already exists');
    } else {
      console.log('User doesn\'t exist');
    }
  });
```

swashcap commented:
Rebased so the merge will be clean.

swashcap commented:
`dist/*` ignore rule dropped in 8e543a24ed2b7dcb453200a04215a059b63c96dc.

swashcap commented:
Closes #20. Closes #26. Closes #40. Closes #43. Closes #44. Closes #43.

jwtlake commented:
Looks like there is an error when running SP tests.
```  
79 passing (19s)
  21 pending
  1 failing

  1) Statistics GET /statistics:
     TypeError: Cannot read property 'get' of undefined
      at Context.it (test/integration/statistics.js:19:51)
```

swashcap commented:
Broken tests were due to merging #95, which created a new integration test file that used the old `apiClient.statistics.get()`-style API. Fixed and rebased.

swashcap commented:
@jwtlake Want to merge this after #107’s merged and tests pass?

jwtlake commented:
🎉

swashcap commented:
:+1: 🍰
